src/System.CommandLine/Command.cs

```diff
@@ -129,59 +166,49 @@ public void AddGlobalOption(Option option)
 
-        private protected void AddSymbol(Symbol symbol)
-        {
-            Children.AddWithoutAliasCollisionCheck(symbol);
-            symbol.AddParent(this);
-        }
-
         /// <inheritdoc />
         public override IEnumerable<CompletionItem> GetCompletions(CompletionContext context)
         {
-            if (Children.Count == 0)
+            List<CompletionItem> completions = new();
+
+            foreach (Command command in SubCommands)
             {
-                return Array.Empty<CompletionItem>();
+                AddCompletions(context.WordToComplete, command, completions);
             }
 
-            var completions = new List<CompletionItem>();
+            foreach (Option option in Options)
+            {
+                AddCompletions(context.WordToComplete, option, completions);
+            }
 
-            if (context.WordToComplete is { } textToMatch)
+            foreach (Argument argument in Arguments)
             {
-                for (var i = 0; i < Children.Count; i++)
+                foreach (var completion in argument.GetCompletions(context))
                 {
-                    var child = Children[i];
-
-                    switch (child)
+                    if (completion.Label.ContainsCaseInsensitive(context.WordToComplete))
                     {
-                        case IdentifierSymbol identifier when !child.IsHidden:
-                            foreach (var alias in identifier.Aliases)
-                            {
-                                if (alias is { } &&
-                                    alias.ContainsCaseInsensitive(textToMatch))
-                                {
-                                    completions.Add(new CompletionItem(alias, CompletionItemKind.Keyword, detail: child.Description));
-                                }
-                            }
-
-                            break;
-
-                        case Argument argument:
-                            foreach (var completion in argument.GetCompletions(context))
-                            {
-                                if (completion.Label.ContainsCaseInsensitive(textToMatch))
-                                {
-                                    completions.Add(completion);
-                                }
-                            }
-
-                            break;
+                        completions.Add(completion);
                     }
                 }
             }
 
             return completions
                    .OrderBy(item => item.SortText.IndexOfCaseInsensitive(context.WordToComplete))
                    .ThenBy(symbol => symbol.Label, StringComparer.OrdinalIgnoreCase);
+
+            static void AddCompletions(string wordToComplete, IdentifierSymbol symbol, List<CompletionItem> completions)
+            {
+                if (!symbol.IsHidden)
+                {
+                    foreach (string alias in symbol.Aliases)
```

Comment:
`Aliases` is currently backed by a `HashSet<string>` which was to avoid collisions but if we're moving this sort of validation to testing, this might be a good candidate for a `List<string>`, faster iteration, and no up-front hash set cost.